File: sdk-java/src/test/java/io/superdurable/dex/integ/InternalChannelTest.java
/*
 * Portions of this file are derived from indeedeng/iwf-java-sdk.
 * Those portions are licensed under the Apache License, Version 2.0.
 * See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
 *
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications are licensed under the Super Durable Source License 1.0.
 * Third-Party Materials remain under the Apache License, Version 2.0.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex.integ;

import io.superdurable.dex.Client;
import io.superdurable.dex.testing.DexDevTestEnvironment;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Path;
import java.time.Duration;
import java.util.Arrays;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;

@Tag("dex-dev")
public final class InternalChannelTest {
    private static final InternalChannelWorkflow WORKFLOW = new InternalChannelWorkflow();
    private static final InternalChannelWaitingWorkflow WAITING_WORKFLOW =
            new InternalChannelWaitingWorkflow();

    @TempDir
    Path cacheDirectory;

    @Test
    void testBasicInternalChannel() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                WORKFLOW)) {
            final String flowId = "basic-internal-" + UUID.randomUUID();
            environment.client().startFlow(WORKFLOW, flowId, 1);
            assertEquals(2, environment.client().waitForFlow(
                    flowId,
                    Integer.class,
                    Duration.ofSeconds(30)));
        }
    }

    @Test
    void testWaitingInternalChannel() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                WAITING_WORKFLOW)) {
            final String flowId = "waiting-internal-" + UUID.randomUUID();
            environment.client().startFlow(WAITING_WORKFLOW, flowId, 1);
            environment.client().publish(
                    flowId,
                    WAITING_WORKFLOW.channel,
                    Arrays.asList(2, 3));
            assertEquals(6, environment.client().waitForFlow(
                    flowId,
                    Integer.class,
                    Duration.ofSeconds(30)));
        }
    }

    void compileBasicInternalChannel(final Client client) {
        client.startFlow(WORKFLOW, "basic-internal", 1);
        final Integer output = client.waitForFlow("basic-internal", Integer.class);
        consume(output);
    }

    void compileWaitingInternalChannel(final Client client) {
        client.startFlow(WAITING_WORKFLOW, "waiting-internal", 1);
        client.publish(
                "waiting-internal",
                WAITING_WORKFLOW.channel,
                Arrays.asList(2, 3));
        final Integer output = client.waitForFlow("waiting-internal", Integer.class);
        consume(output);
    }

    private static void consume(final Object value) {
    }
}

File: sdk-typescript/test/iwfcompat/internal_channel_test.ts
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Super Durable Source License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

import { doubleCodec, type Client } from "../../src/index.js";

import * as flows from "./iwf_flows.js";

export async function compileBasicInternalChannel(client: Client): Promise<void> {
  await client.startFlow(flows.BASIC_INTERNAL, "basic-internal", 1);
  const output: number = await client.waitForFlow("basic-internal", doubleCodec);
  void output;
}

export async function compileWaitingInternalChannel(client: Client): Promise<void> {
  const flow = flows.WAITING_INTERNAL;
  await client.startFlow(flow, "waiting-internal", 1);
  await client.publish("waiting-internal", flow.channel, 2, 3);
  const output: number = await client.waitForFlow("waiting-internal", doubleCodec);
  void output;
}
